Implement local dynamic filtering for broadcast inner-joins

rzeyde-varada opened the pull request:
# Overview 
When a broadcast inner-join is used, probe-side table scan and hash build operators are running in the same process - so the dynamic filter can be extracted and applied locally, without involving the coordinator.
The connector may optimize the probe-side scan significantly by applying the dynamic filter using run-time predicate pushdown (derived from the build-side collected values).

We have implemented and tested the suggested API for [our Presto connector](https://www.slideshare.net/OriReshef/presto-for-apps-deck-varada-prestoconf) - and it significantly improves the performance of highly-selective joins, e.g. [TPCDS q71](https://github.com/prestosql/presto/blob/master/presto-product-tests/src/main/resources/sql-tests/testcases/tpcds/q71.sql) took ~1.5 minutes without dynamic filtering, and ~13 seconds with dynamic filtering.
For more details, see [the following slides](https://www.slideshare.net/OriReshef/dynamic-filtering-for-presto-join-optimisation) and #52 (roadmap tracking issue).

# Open issues and questions/notes to reviewers

## Naming
Since #91 is already merged, there is a possible confusion between the methods.
Both of them use dynamic filtering approach, but this PR enables it only for broadcast inner-joins and doesn't require any changes to the optimizer and the coordinator.
Please let us know about the preferred naming for this feature (i.e. its class names and session parameters).

## Plan parsing
The existing code supports some intermediate plan nodes between the join node and the probe scan node (see [here](https://github.com/prestosql/presto/pull/931/files#diff-d3d099fe2c1d2a89408b9234c0d7ab29R48)). It is possible to extend it to support more plan node types (e.g. FilterNode).

## Integration tests
Currently there are none, since we are using an external CI framework for testing this feature.
It’s possible to adapt an existing connector to support the new API (e.g. [MemoryConnector](https://github.com/rzeyde-varada/prestosql/commit/d6a409c097)).

## Abstracting TupleDomain
Currently we collect and apply the dynamic filter as a `TupleDomain` (for simplicity reasons).
It is possible to abstract it into a Connector-provided abstract data type (i.e. `ConnectorDynamicFilter`) - which can be used by connectors to support bloom filters.
What do you think?

martint commented:
@raunaqmorarka, let's figure out how this fits in with the work you've been doing.

sopel39 commented:
Hi @rzeyde-varada 

Thanks for the PR. This is a neat change! I think we should coordinate your solution with dynamic filtering feature that is in progress (cc @raunaqmorarka) as @martint  mentioned.
In the dynamic filtering that is in progress, dynamic filters will be collected by coordinator and then sent back to all workers (I think that will be in next dynamic filters PR).
In case of broadcast join and source stage we could still send tuple domains to coordinator (so that we can filter entire partitions) but also short-circut directly to scan operator.

I like how you push down dynamic filter tuple domains to page source. I think the original idea (behind dynamic filtering that is being landed) is to evaluate rows on tuple domains in scan operator itself. However, I think your approach has additional advantages as connector can filter files using file header metadata (e.g ORC). Filtering in scan operator is still beneficial. It provides lower granularity and entire pages could be filtered out. Late materialization will also cover row filtering in source stage broadcast join.

> Since #91 is already merged, there is a possible confusion between the methods.

I think both solutions can be integrated so there is no need for splitting between different "dynamic filter" types. Please take a look at dynamic filters execution placeholders in `LocalExecutionPlanner`. Maybe you could use them to incorporate your short-circut execution implementation?

> It’s possible to adapt an existing connector to support the new API

I think you could use Hive connector and collect input statistics from table scan operator. You could then check that stats are zero when dynamic filtering is used and works.

> It is possible to abstract it into a Connector-provided abstract data type (i.e. ConnectorDynamicFilter) - which can be used by connectors to support bloom filters.

Could you provide an example of connector which could take advantage of this? Explicit row filtering (using bloom filters) could be performed in Presto engine itself.
Usually file formats (e.g ORC) provide bloom-filters that can be used to check for potential value existence.

rzeyde-varada commented:
Thanks for your kind works @sopel39!

> Maybe you could use them to implement your "short-circuit" filtering?

Sounds good - I can move our implementation there.

> You could then check that stats are zero when dynamic filtering is used and works.

I have followed your advice and added a simple dynamic filtering implementation and test to the Memory connector (since it's simpler than Hive connector) here: https://github.com/rzeyde-varada/prestosql/commit/d6a409c097bc2df86773f3151eb44b9491438c0d.
What do you think?

> Could you provide an example of connector which could take advantage of this? Explicit row filtering (using bloom filters) could be performed in Presto engine itself.

Currently, I am not aware of any connectors that support bloom-filtering. 
The idea behind the suggestion was that we've noticed that TupleDomain is used both for predicate pushdown and dynamic filtering. However, dynamic filtering may have a small false-positive error rate, so we could construct a bloom-filter during the collection phase, and pass it to the page source provider: https://github.com/prestosql/presto/pull/931/commits/eaf2be4ae6e9c21fd316ea175e899299f9aa3356#diff-3bb2ddad9db7856c6a9ce4a76dcd81aaR29.

I agree that we can postpone the discussion to a later PR.

findepi commented:
> Currently, I am not aware of any connectors that support bloom-filtering.

@rzeyde-varada is https://github.com/prestosql/presto/pull/914 > `Add support for ORC UTF8 bloom filters` related?

rzeyde-varada commented:
Thanks for the heads-up! Will take a look :)

rzeyde-varada commented:
> Thanks for the heads-up! Will take a look :)

I was thinking on a bit different approach (IIUC the PR):
- During the collection phase of the join, we'll construct a bloom filter using the values coming from the build-side table (instead of a TupleDomain). 
- The bloom filter will be propagated to the probe-side page source providers, and will be used to check whether there is an [intersection](https://en.wikipedia.org/wiki/Bloom_filter#The_union_and_intersection_of_sets) with the bloom filter of each page from the storage layer.

It's a bit similar to how the filtering is done by [TupleDomainOrcPredicate.java](https://github.com/prestosql/presto/blob/74ace71f60e8cb9d9c41d9d95a05002b9a72e801/presto-orc/src/main/java/io/prestosql/orc/TupleDomainOrcPredicate.java#L123), but by using a bloom filter instead of a discrete value list, we can hopefully support larger dynamic filters (following the [design document](https://docs.google.com/document/d/1TOlxS8ZAXSIHR5ftHbPsgUkuUA-ky-odwmPdZARJrUQ/edit#heading=h.3plqme3qiwts)).

dain commented:
> * During the collection phase of the join, we'll construct a bloom filter using the values coming from the build-side table (instead of a TupleDomain).
> * The bloom filter will be propagated to the probe-side page source providers, and will be used to check whether there is an [intersection](https://en.wikipedia.org/wiki/Bloom_filter#Interesting_properties) with the bloom filter of each page from the storage layer.
> 
> It's a bit similar to how the filtering is done by [TupleDomainOrcPredicate.java](https://github.com/prestosql/presto/blob/74ace71f60e8cb9d9c41d9d95a05002b9a72e801/presto-orc/src/main/java/io/prestosql/orc/TupleDomainOrcPredicate.java#L123), but by using a bloom filter instead of a discrete value list, we can hopefully support larger dynamic filters (following the [design document](https://docs.google.com/document/d/1TOlxS8ZAXSIHR5ftHbPsgUkuUA-ky-odwmPdZARJrUQ/edit#heading=h.3plqme3qiwts)).

I don't think pushing a bloom filter into connectors will work at all.  The problem is a bloom filter requires an exact implementation specific configuration of things like hash algorithm, number of hashes, false positive rate, and even stuff as subtle as how values are encoded into bytes before hashing.  For Hive alone it is a big pain, because ORC already has two different bloom filter encodings, and I think Parquet has a different system.

The good news is for this local node dynamic pushdown, we actually have all of the values in memory, so we could simply pass the pages to the connector and it could build any data structures they want.

rzeyde-varada commented:
> The good news is for this local node dynamic pushdown, we actually have all of the values in memory, so we could simply pass the pages to the connector and it could build any data structures they want.

You're right, thanks for the correction!

rzeyde-varada commented:
I have updated the PR to use the existing `enable_dynamic_filtering` session property, and added a simple integration test at [`TestMemorySmoke.java`](https://github.com/prestosql/presto/pull/931/commits/64d1c61e3211e4c0019fa10d38da46442ebcf080#diff-6102dfcf993241d1413b4fc63e4ca5a9). 
Could you please take a look?

raunaqmorarka commented:
@sopel39 @rzeyde-varada I think there is value in putting this local dynamic filtering implementation behind a separate flag. It will allow easier testing and rollout of this feature as well as the changes I'm working on. How about we always enable this feature if `experimental.enable-dynamic-filtering` is enabled but allow disabling it through a separate flag `disable-local-dynamic-filtering` ?

rzeyde-varada commented:
> How about we always enable this feature if `experimental.enable-dynamic-filtering` is enabled but allow disabling it through a separate flag `disable-local-dynamic-filtering` ?

Sounds good to me :)

raunaqmorarka commented on presto-main/src/main/java/io/prestosql/operator/CollectOperator.java:
```diff
@@ -0,0 +1,151 @@
+        this.context = requireNonNull(context, "context is null");
+        this.predicateCollector = dynamicFilter.collector();
+        this.dynamicPredicateFuture = dynamicPredicateFuture;
+        this.remainingRowCount = getDynamicFilteringMaxRowCount(context.getSession());
```
I think this will result in the behavior that if any driver on build side sees more than `dynamic_filtering_max_row_count` rows then we give up. Should we rename the config to `dynamic_filtering_max_per_driver_row_count` to avoid confusion ?

rzeyde-varada replied:
Sounds good, changed in 4dd335f.

raunaqmorarka commented on presto-main/src/main/java/io/prestosql/operator/CollectOperator.java:
```diff
@@ -0,0 +1,151 @@
+        if (dynamicPredicateFuture.isCancelled()) {
+            return;  // the predicate became too large.
+        }
+        if (remainingRowCount >= page.getPositionCount()) {
```
Apart from a limit on number of positions, in the older implementation we had also used a limit on size for the case where join column is a long VARCHAR
https://github.com/prestodb/presto/pull/9453#issuecomment-353018791

rzeyde-varada replied:
Good idea, thanks!
Added in 4dd335f.

raunaqmorarka commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,261 @@
+        {
+            // Collect only the columns which are relevant for the JOIN.
+            Page projectedPage = selector.apply(page);
+            positionsCount += projectedPage.getPositionCount();
```
We would want to count only non-null values in `positionsCount` and `sizeInBytes` as null values won't contribute to creating a large predicate. I don't think current code takes care of that case.

rzeyde-varada replied:
Good idea, thanks!
Please take a look at https://github.com/rzeyde-varada/prestosql/commit/edb2d36b90517b5601136e56c8f11ef835a56e97 (pushed in a separate [wip/dynamic-filter](https://github.com/rzeyde-varada/prestosql/commits/wip/dynamic-filter) branch).

A small question about `Block#getEstimatedDataSizeForStats` - I have used this method for estimating the byte size of each value in order to compute `sizeInBytes` for dynamic filter collection. Is it OK?

raunaqmorarka replied:
I think it should be ok, an alternative is to use `value.toString().getBytes().length`, though it looks like `getEstimatedDataSizeForStats` might be the more straightforward way.
You might want to check the performance impact of using HashSet to get distinct values count in rzeyde-varada@edb2d36. If the overhead is minimal then getting distinct values is good, we had avoided it earlier to keep this part as lightweight as possible.

rzeyde-varada replied:
> You might want to check the performance impact of using HashSet to get distinct values count in [rzeyde-varada/prestosql@edb2d36](https://github.com/rzeyde-varada/prestosql/commit/edb2d36). 

Thanks, will add a benchmark.

rzeyde-varada replied:
Added a simple benchmark, rebased and force-pushed 733ffaf10c.
When running on my machine, the overhead seems to be quite low:

With `collectOperator` running after `tableScanOperator`:
```
dynamic_filter_collect ::   20.360 cpu ms ::    0B peak memory :: in   15K,      0B,    737K/s,      0B/s :: out   15K,   132KB,    737K/s,  6.32MB/s
```

Without `collectOperator` (only `tableScanOperator`):
```
dynamic_filter_collect ::   19.855 cpu ms ::    0B peak memory :: in   15K,      0B,    755K/s,      0B/s :: out   15K,   132KB,    755K/s,  6.48MB/s
```

raunaqmorarka replied:
LGTM

rzeyde-varada replied:
Thanks, will cherry-pick https://github.com/prestosql/presto/commit/733ffaf10ccd86bc1872d90ed54d2440c3fbad2b into `dynamic-filter` branch.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/CollectOperator.java:
```diff
@@ -0,0 +1,151 @@
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxRowCount;
+import static java.util.Objects.requireNonNull;
+
+/// This operator acts as a simple "pass-through" pipe, while saving its input pages.
```
use `/* ... */` style of comments

rzeyde-varada replied:
Fixed at 819e419e8a

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/CollectOperator.java:
```diff
@@ -0,0 +1,151 @@
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.dynamicFilter = dynamicFilter;
```
Add `requireNonNull` check for `dynamicFilter`

rzeyde-varada replied:
Fixed at 819e419

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/CollectOperator.java:
```diff
@@ -0,0 +1,151 @@
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            CompletableFuture<TupleDomain<ColumnHandle>> dynamicPredicateFuture = new CompletableFuture<>();
```
use Guava `ListenableFutures` as it's used in other places in `presto-main`. We only use `CompletableFuture` for spi

rzeyde-varada replied:
Fixed at 819e419

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/CollectOperator.java:
```diff
@@ -0,0 +1,151 @@
+            dynamicFilter.addFuture(dynamicPredicateFuture);
+            return new CollectOperator(
+                    driverContext.addOperatorContext(operatorId, planNodeId, CollectOperator.class.getSimpleName()),
+                    dynamicFilter, dynamicPredicateFuture);
```
each argument in newline

rzeyde-varada replied:
Fixed at 819e419

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/CollectOperator.java:
```diff
@@ -0,0 +1,151 @@
+/// This operator acts as a simple "pass-through" pipe, while saving its input pages.
+/// The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+/// We support only small build-side pages (which should be the case when using "broadcast" join).
+public class CollectOperator
```
I think `DynamicFilterCollectOperator` would be better as it's more explicit

rzeyde-varada replied:
Fixed at 819e419

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1935,14 +1956,26 @@ private PhysicalOperation createLookupJoin(JoinNode node,
                 LocalExecutionPlanContext context,
-                boolean spillEnabled)
+                boolean spillEnabled,
+                Optional<DynamicFilter> dynamicFilter)
```
please use `io.prestosql.sql.planner.plan.JoinNode#getDynamicFilters` instead.

rzeyde-varada replied:
If I understood correctly, it should be done at 39a2eb47da.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1915,13 +1932,17 @@ private PhysicalOperation createLookupJoin(JoinNode node,
         {
-            // Plan probe
-            PhysicalOperation probeSource = probeNode.accept(this, context);
+            Optional<DynamicFilter> dynamicFilter = createDynamicFilter(node, probeNode, probeSymbols, buildNode, buildSymbols, context);
```
please plug into
```
            // TODO: Execution must be plugged in here
            if (!node.getDynamicFilters().isEmpty()) {
                log.debug("[Join] Dynamic filters: %s", node.getDynamicFilters());
            }
```
within `visitJoin` code

rzeyde-varada replied:
I moved it to `createLookupSourceFactory`, following https://github.com/prestosql/presto/pull/931#discussion_r299817744.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1915,13 +1932,17 @@ private PhysicalOperation createLookupJoin(JoinNode node,
         {
-            // Plan probe
-            PhysicalOperation probeSource = probeNode.accept(this, context);
+            Optional<DynamicFilter> dynamicFilter = createDynamicFilter(node, probeNode, probeSymbols, buildNode, buildSymbols, context);
```
If dynamic filter would not be consumed by table scan then it should not be created. You should check that plan fragment contains table scans (and validate that it's for this particular dynamic filter within join)

rzeyde-varada replied:
Done at 39a2eb47da.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,244 @@
+        return Optional.empty();
+    }
+
+    public static Optional<DynamicFilter> create(JoinNode node, PlanNode probeNode, List<Symbol> probeSymbols, PlanNode buildNode, List<Symbol> buildSymbols, TypeProvider types)
```
In current master `JoinNode` already contains dynamic filters `io.prestosql.sql.planner.plan.JoinNode#getDynamicFilters` and table scans contain expressions that are "dynamic filters".
Planning part is done, so I think we should reuse it.

rzeyde-varada replied:
Good point, thanks!
Done at 39a2eb47da.

rzeyde-varada commented:
Many thanks for the review!
Will update the PR in the following days.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/CollectOperator.java:
```diff
@@ -0,0 +1,151 @@
+    private boolean finished;
+    private Page current;
+    private final DynamicFilter.Collector predicateCollector;
+    private final CompletableFuture<TupleDomain<ColumnHandle>> dynamicPredicateFuture;
```
I think instead of `CompletableFuture<TupleDomain<ColumnHandle>>` (or `SettableFuture`) `DynamicCollector` could just be passed `Consumer<TupleDomain<ColumnHandle>>`.
It seems that cancellation doesn't really matter for consumer of `TupleDomain` predicates

rzeyde-varada replied:
Good catch, thanks!
Done at 39a2eb47da.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    /// Must be called after all futures have been added, i.e. by CollectOperatorFactory::noMoreOperators() method.
+    public void noMoreFutures()
```
This shouldn't be necessary, because the number of futures is know at planning time.
It's `buildContext.getDriverInstanceCount().orElse(1);`  (see `createLookupSourceFactory`)

rzeyde-varada replied:
Thanks, done at 39a2eb47da.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    /// Each CollectOperator should submit its own predicate into a separate future (one for each partition).
+    public void addFuture(CompletableFuture<TupleDomain<ColumnHandle>> future)
```
You can make it other way around. That is `DynamicFilter` could provide `Consumer<TupleDomain<ColumnHandle>>` and consume predefined number of tuple domains (the number is known).

rzeyde-varada replied:
Sounds good, done at 39a2eb47da.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,244 @@
+        });
+    }
+
+    public class Collector
```
Inline `Collector` into `CollectOperator` so it's logic is contained within single class

rzeyde-varada replied:
Done at 39a2eb47da.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,244 @@
+    public class Collector
+    {
+        private final ImmutableList.Builder<Page> pageBuilder;
+        private final PageChannelSelector selector;
```
remaining rows logic should be part of `Collector`, but once you inline it it should not matter.

rzeyde-varada replied:
Done at 39a2eb47da.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,244 @@
+        public void add(Page page)
+        {
+            // Collect only the columns which are relevant for the JOIN.
+            pageBuilder.add(selector.apply(page));
```
once the remaining rows is exceeded, you could drop references to collected pages

rzeyde-varada replied:
Done at 39a2eb47da.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    /// Decorate existing page source provider, to allow dynamic filtering.
+    public PageSourceProvider wrap(PageSourceProvider pageSourceProvider)
```
I think it would be cleaner (instead of `PageSourceProvider` wrapper) if `DynamicFilter` provided `ListenableFuture` that provides aggregated `TupleDomain`. Then one could add listener that would `applyDynamicFilter` to `pageSourceProvider`.

rzeyde-varada replied:
Done at 39a2eb47da (but I still had to create a "dynamic" page source provider to poll for dynamic filters during the execution of the operator over the table splits - see the `createDynamicFilteringPageSourceProvider` method).

sopel39 replied:
> but I still had to create a "dynamic" page source provider to poll for dynamic filters during the execution of the operator over the table splits - see the createDynamicFilteringPageSourceProvider method

Why so? You have access to `PageSource` in `LocalExecutionPlanner`. Why not call `applyDynamicFilters` to it when `TupleDomain` future gets done? I think `ConnectorPageSource` is guaranteed to be thread safe.

rzeyde-varada replied:
IIUC, the issue is that the new `PageSourceProvider#applyDynamicFilter` API operates on a `TableHandle`, returning potentially new and filtered `TableHandle` - similarly to how it's done be `Metadata#applyFilter` API.
https://github.com/prestosql/presto/pull/931/files#diff-3bb2ddad9db7856c6a9ce4a76dcd81aaR29
An example implementation for `MemoryPageSourceProvider`:
https://github.com/prestosql/presto/pull/931/files#diff-8b4a37d650086d97552712b506dfb6a8R85

Initially, I have intended that `PageSourceProvider#applyDynamicFilter` would be called just before creating a new `PageSource` for each split with the current dynamic filter (if available) - "restricting" the effective predicate on the `TableHandle` to be used for the current split.
I did it by "wrapping" the existing `PageSourceProvider` to keep the required changes to the `LocalExecutionPlanner` module.

I am not sure that I understand your suggestion above, so please let me know if we need to change some of the existing APIs to make the overall design better.

sopel39 replied:
Ok. That makes sense. Thanks!

sopel39 commented:
Could you squash fixups?

rzeyde-varada commented:
> Could you squash fixups?

Done.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1935,6 +1978,37 @@ private PhysicalOperation createLookupJoin(JoinNode node,
+            }
+            return DynamicFilter
+                    .create(node, context.getTypes())
+                    .map(filter -> {
```
You can use `MoreFutures#addSuccessCallback`.
This can be then simplified to:
```
filter -> addSuccessCallback(filter.getResultFuture(), tuple -> context.intersectDynamicFilter(filter.getTableScanNode(), tuple))
```
assuming future does not return optional and errors are reported and handled in `DynamicFilter`.

sopel39 replied:
Actually. Scratch that.
I think you could simply check if `filter.getResultFuture()` is done when you create new `PageSource` for particular table scan.

rzeyde-varada replied:
> You can use `MoreFutures#addSuccessCallback`.

Thanks, done at 7b91508809.

rzeyde-varada replied:
> I think you could simply check if filter.getResultFuture() is done when you create new PageSource for particular table scan.

This can work, but in order to support multiple join nodes in the same plan fragment (e.g. [TPCDS q13](https://github.com/prestosql/presto/blob/master/presto-product-tests/src/main/resources/sql-tests/testcases/tpcds/q13.sql)), we'll need to intersect several future results for the same TableScan node. 
I have added 972c66d15f for supporting this use-case, WDYT?

sopel39 replied:
> Can we assume that there is a single TableScan in each stage?

No. In colocated join there might be two table scans for different tables in same scan.

rzeyde-varada replied:
Good point, thanks for the clarification!

rzeyde-varada commented:
Force-pushed a small fix to the PageSourceProvider `applyDynamicFilter` API (adding `ConnectorTransactionHandle` argument, similar to the `createPageSource` API).

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,170 @@
+        return tableScanNode;
+    }
+
+    public ListenableFuture<Optional<TupleDomain<ColumnHandle>>> getResultFuture()
```
Good point, fixed at 7b91508809.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1935,6 +1978,37 @@ private PhysicalOperation createLookupJoin(JoinNode node,
+                            @Override
+                            public void onFailure(Throwable t)
+                            {
+                                log.error(t, "Dynamic filtering collection future failed");
```
Since I am using a `SettableFuture` and don't call its `setException` method, I think that we may not need to handle future failures.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,229 @@
+        }
+
+        @Override
+        public void accept(Optional<TupleDomain<ColumnHandle>> dynamicPredicate)
```
I don't think it's needed. Could you add a method like:
```
Consumer<TupleDomain<ColumnHandle>> DynamicFilter#getPartitionTupleDomainConsumer()
```
On last consumed partition it would do the union and set final future with `TupleDomain` for all partitions.

rzeyde-varada replied:
Sounds good, fixed at c343902c15.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,229 @@
+        private final DynamicFilter dynamicFilter;
+
+        private TupleDomain<ColumnHandle>[] partitions;
+        private int partitionCount;
```
You can just use `AtomicInteger` to do the count. Then you don't need `partitionLock` anymore.

rzeyde-varada replied:
Yes, this is indeed much better -> c343902c15.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,229 @@
+                    TupleDomain<ColumnHandle> dynamicConstraint = Arrays.stream(partitions).reduce(TupleDomain.none(), TupleDomain::columnWiseUnion);
+                    // NOTE: may result in a bit more relaxed constraint if there are multiple columns and multiple rows.
+                    // See the comment at TupleDomain::columnWiseUnion() for more details.
+                    dynamicFilter.accept(dynamicConstraint);
```
callback should not be called within critical section, see: `io.prestosql.operator.exchange.LocalExchangeSource#addPage`

rzeyde-varada replied:
Should be fixed by resolving https://github.com/prestosql/presto/pull/931#discussion_r302104425 at c343902c15.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,229 @@
+
+        if (positionsCount > maxRowCount || sizeInBytes > maxByteSize) {
+            // The resulting predicate is too large, allow all probe-side values to be read.
+            dynamicPredicateConsumer.accept(Optional.empty());
```
Why is this needed? I think we can simply skip calling `accept` with `Optional.empty()`

rzeyde-varada replied:
Passing an `Optional.empty()` here helps the consumer to know that the collection failed early, so it would drop previously collected predicates:
https://github.com/prestosql/presto/blob/c343902c156401496bc89e6ca671a9747842594c/presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java#L66

rzeyde-varada replied:
On the other hand, we can pass `TupleDomain.all()` (similar to how it's done by #1072).
Should be fixed at 1559eeb7b4.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,202 @@
+
+        this.partitions = new TupleDomain[partitionCount];
+        this.partitionCounter = new AtomicInteger();
+        this.tupleDomainConsumer = tupleDomain -> {
```
extract that to separate method outside of constructor

rzeyde-varada replied:
Done at https://github.com/prestosql/presto/commit/96a591f3652533cd2e2795d349556158e2b86b6b#diff-d3d099fe2c1d2a89408b9234c0d7ab29R99.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,202 @@
+            final int partitionIndex = this.partitionCounter.getAndIncrement();
+            verify(partitionIndex < partitions.length);
+            partitions[partitionIndex] = tupleDomain;
+            if (partitionIndex == partitions.length - 1) {
```
I don't think that this check is sufficient to ensure that all partitions are filled. Since `partitionCounter` is updated before `partitions` are filled there is no guarantee that during this check all `partitions` are filled.

I think some synchronization is required here.

rzeyde-varada replied:
Should be fixed here: https://github.com/prestosql/presto/commit/96a591f3652533cd2e2795d349556158e2b86b6b#diff-d3d099fe2c1d2a89408b9234c0d7ab29R102

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,191 @@
+    private final int maxValueCount;
+    private final long maxByteSize;
+
+    private final List<DynamicFilter.Descriptor> descriptors;
```
please operate on `io.prestosql.sql.DynamicFilters.Descriptor` instead since it's technically not required to filter at `ColumnHandle` level. This way we could reuse the operator for collecting dynamic filters for partition filtering in coordinator

rzeyde-varada replied:
I switched to use `TupleDomain<Symbol>` almost everywhere, except at `LocalExecutionPlanner#LocalExecutionContext#getDynamicFilter()`.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,202 @@
+        PlanNode buildNode = planNode.getRight();
+
+        Map<String, Symbol> filterSymbols = new HashMap<>(); // mapping filter ID to its probe-side symbol
+        while (currentNode != null) {
```
you can use `io.prestosql.sql.planner.optimizations.PlanNodeSearcher` to search for all table scans with filter on top (which have matching dynamic filters)

rzeyde-varada replied:
https://github.com/prestosql/presto/commit/4561c62d7d5b26a2fa12c623d4f279daac7360ea#diff-d3d099fe2c1d2a89408b9234c0d7ab29R127

sopel39 commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorPageSourceProvider.java:
```diff
@@ -42,4 +44,14 @@ default ConnectorPageSource createPageSource(
         throw new UnsupportedOperationException("createPageSource() must be implemented");
     }
+
+    default ConnectorTableHandle applyDynamicFilter(
```
I wonder if it just wouldn't be simpler to pass dynamic filter directly to `createPageSource`. `ConnectorTableHandle` feels more like a planning object, but in `LocalExecutionPlanner` we are already in execution part.

rzeyde-varada replied:
Sounds good, will add a new `createPageSource` overload (accepting the dynamic filter as the last argument), IIUC.

rzeyde-varada replied:
https://github.com/prestosql/presto/commit/4561c62d7d5b26a2fa12c623d4f279daac7360ea#diff-4dc65ea280f82471ef8da2070c3b9768

rzeyde-varada commented:
Rebased, squashed and force-pushed https://github.com/prestosql/presto/commit/4561c62d7d5b26a2fa12c623d4f279daac7360ea.

sopel39 commented on presto-benchmark/src/main/java/io/prestosql/benchmark/AbstractOperatorBenchmark.java:
```diff
@@ -181,7 +181,7 @@ protected final OperatorFactory createTableScanOperator(int operatorId, PlanNode
             {
                 OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, planNodeId, "BenchmarkSource");
-                ConnectorPageSource pageSource = localQueryRunner.getPageSourceManager().createPageSource(session, split, tableHandle, columnHandles);
+                ConnectorPageSource pageSource = localQueryRunner.getPageSourceManager().createPageSource(session, split, tableHandle, columnHandles, null);
```
I think you can pass tuple domain all here

rzeyde-varada replied:
Done.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,206 @@
+{
+    public static class Channel
+    {
+        private final Type type;
```
The type information is redundant. In `DynamicFilterCollectOperator` you can easily obtain types from `LocalExecutionPlanner`

rzeyde-varada replied:
Thanks, fixed.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,206 @@
+    public static class Channel
+    {
+        private final Type type;
+        private final int buildIndex;
```
rename to `buildChannelIndex`

rzeyde-varada replied:
Done.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,206 @@
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Verify.verify;
+
+public class DynamicFilter
```
I think the name is too close to `DynamicFilters`. Rename it to `LocalDynamicFilter`?

rzeyde-varada replied:
Sounds good, fixed.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,206 @@
+            if (partitions == null) {
+                return;
+            }
+            if (tupleDomain.isAll()) {  // disable dynamic filtering - all build-side values are needed.
```
put comments in separate lines

rzeyde-varada replied:
Done.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,206 @@
+        List<FilterNode> filterNodes = PlanNodeSearcher
+                .searchFrom(planNode.getLeft())
+                .where(DynamicFilter::isFilterAboveTableScan)
+                .recurseOnlyWhen(DynamicFilter::isInnerReplicatedJoin)
```
Why? I don't think extra intermittent nodes are problem as long as filters match dynamic symbols. Dynamic filter planning takes care of proper planning.

rzeyde-varada replied:
Thanks, fixed.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java:
```diff
@@ -0,0 +1,206 @@
+            DynamicFilters.ExtractResult extractResult = DynamicFilters.extractDynamicFilters(filterNode.getPredicate());
+            for (DynamicFilters.Descriptor descriptor : extractResult.getDynamicConjuncts()) {
+                Symbol probeSymbol = Symbol.from(descriptor.getInput());
+                if (!tableScanNode.getAssignments().containsKey(probeSymbol)) {
```
that shouldn't happen. Otherwise dynamic filtering planning would be incorrect

rzeyde-varada replied:
You're right, fixed.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -521,6 +525,9 @@ private static void addLookupOuterDrivers(LocalExecutionPlanContext context)
         private final List<DriverFactory> driverFactories;
         private final Optional<IndexSourceContext> indexSourceContext;
 
+        private TupleDomain<DynamicFilters.Descriptor> dynamicFilter = TupleDomain.all();
```
I think dynamic filter should be passed to subcontext, e.g. in collocated joins there could be two scans of same table (self join)

sopel39 replied:
`TupleDomain<DynamicFilters.Descriptor> dynamicFilter` is not longer a planning object. I think logic from context should be extracted to some `LocalDynamicFiltersCollector` that would contain both `intersectDynamicFilter` and `getDynamicFilterSupplier` methods.

rzeyde-varada replied:
Extracted the functionality into a separate class, allowing subcontexts to use the it as well.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -521,6 +525,9 @@ private static void addLookupOuterDrivers(LocalExecutionPlanContext context)
         private final List<DriverFactory> driverFactories;
         private final Optional<IndexSourceContext> indexSourceContext;
 
+        private TupleDomain<DynamicFilters.Descriptor> dynamicFilter = TupleDomain.all();
```
nit: add a comment that this potentially contains domains for dynamic filters for different table scans (in case of colocated joins)

rzeyde-varada replied:
Done (at `LocalDynamicFiltersCollector`).

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -635,6 +642,28 @@ public void setDriverInstanceCount(int driverInstanceCount)
+                return dynamicFilter
+                        .transform(DynamicFilters.Descriptor::getInput)
+                        .transform(Symbol::from)
+                        .transform(assignments::get);
```
`assignments::get` could return null potentially

rzeyde-varada replied:
IIUC, it should be fine since `TupleDomain#transform` should skip the symbols that doesn't have a corresponding column in `assignments`:
https://github.com/prestosql/presto/blob/bfac56d6b34ec488a3ff132055fc0aca25d5448e/presto-spi/src/main/java/io/prestosql/spi/predicate/TupleDomain.java#L361-L383

rzeyde-varada commented:
Thanks for the review, will address the issues in the beginning of the next week.

rzeyde-varada commented:
Force-pushed 15069754ab (following https://github.com/prestosql/presto/pull/931#pullrequestreview-263554237).

rzeyde-varada commented:
Fixed tests' errors on TPCH connectors (which don't support the new overloaded `ConnectorPageSourceProvider#createPageSource` API) and force-pushed https://github.com/prestosql/presto/commit/aa68369ebbc9eadf23cd149f6fa375118a0976d9.

rzeyde-varada commented:
Fixed a test where there are local dynamic filter should not be applied (force-pushed a42e894428)

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    TupleDomain<ColumnHandle> get(TableScanNode tableScan)
```
just make the whole method `synchronized`

rzeyde-varada replied:
Done.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java:
```diff
@@ -0,0 +1,57 @@
+        this.predicate = TupleDomain.all();
+    }
+
+    void intersect(TupleDomain<DynamicFilters.Descriptor> predicate)
```
just make the whole method `synchronized`

rzeyde-varada replied:
Done.

rzeyde-varada commented:
> Still needs to do final pass on code, but it seems to look good.

Thanks!
Fixed and force-pushed https://github.com/prestosql/presto/commit/7024cc353375859d9fd0992fbd964514617bac54.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+        this.channels = dynamicFilter
+                .getChannels()
+                .stream()
+                .filter(channel -> channel.getDescriptor().getInput() instanceof SymbolReference)
```
That line
`channel.getDescriptor().getInput() instanceof SymbolReference`
doesn't seem to be valid.
`Descrptor` is probe side specific. Why would `DynamicFilterCollectOperator` care if it's a symbol reference or not?

rzeyde-varada replied:
Good point, thanks! 
Changed `DynamicFilterCollectOperator` so it won't have to work with `Descriptor`s.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+    private final OperatorContext context;
+    private boolean finished;
+    private Page current;
+    private final Consumer<TupleDomain<DynamicFilters.Descriptor>> dynamicPredicateConsumer;
```
I think it should be `Consumer<TupleDomain<String>>` where `String` is key from `io.prestosql.sql.planner.plan.JoinNode#dynamicFilters`

`DynamicFilters.Descriptor` is specific to probe side

rzeyde-varada replied:
Good point, thanks! 
Changed `DynamicFilterCollectOperator` so it won't have to work with `Descriptor`s.

sopel39 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java:
```diff
@@ -83,6 +83,12 @@ public HivePageSourceProvider(
 
     @Override
     public ConnectorPageSource createPageSource(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorSplit split, ConnectorTableHandle table, List<ColumnHandle> columns)
```
I think `TestHiveDistributedJoinQueries` should be tested with dynamic filtering on and off

sopel39 replied:
Also you could add tests that ensure that with dynamic filtering less input data is read, see `TestWorkProcessorPipelineQueries` for example

rzeyde-varada replied:
Sounds good, added `TestHiveDistributedJoinQueriesWithDynamicFiltering`.

rzeyde-varada commented on presto-benchmark/src/main/java/io/prestosql/benchmark/DynamicFilterCollectBenchmark.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @Override
+    protected List<? extends OperatorFactory> createOperatorFactories()
```
Thanks, added `BenchmarkDynamicFilterCollectOperator`.
Please let me know if it can be improved.

rzeyde-varada replied:
Removed.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+
+    private final List<LocalDynamicFilter.Channel> channels;
+    private final Type[] buildTypes;
+    private HashSet<Object>[] valueSets;  // may be dropped if the predicate becomes too large.
```
you can use `io.prestosql.operator.aggregation.TypedSet` instead

rzeyde-varada replied:
Thanks for the suggestion!
Refactored to use `TypedSet` and a separate `BlockBuilder` for each channel.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+        private final LocalDynamicFilter dynamicFilter;
+        private final Type[] buildTypes;
+
+        public DynamicFilterCollectOperatorFactory(int operatorId, PlanNodeId planNodeId, LocalDynamicFilter dynamicFilter, List<Symbol> buildSymbols, TypeProvider types)
```
I would rather make `DynamicFilterCollectOperator` not dependent on `LocalDynamicFilter`. This way the operator can be reused for non-local dynamic filter too

rzeyde-varada replied:
You're right, refactored the operator so it won't depend on `LocalDynamicFilter `.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+        }
+
+        // Collect only the columns which are relevant for the JOIN.
+        int index = 0;
```
Done.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+    private final List<LocalDynamicFilter.Channel> channels;
+    private final Type[] buildTypes;
+    private HashSet<Object>[] valueSets;  // may be dropped if the predicate becomes too large.
+    private boolean[] nullAllowed;
```
Sounds good, removed.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+    private boolean[] nullAllowed;
+
+    private long valuesCount;
+    private long sizeInBytes;
```
Done.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+    private HashSet<Object>[] valueSets;  // may be dropped if the predicate becomes too large.
+    private boolean[] nullAllowed;
+
+    private long valuesCount;
```
Done.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+
+        this.buildTypes = requireNonNull(buildTypes, "buildTypes is null");
+
+        this.valueSets = channels.stream().map(d -> new HashSet<>())
```
Fixed.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,207 @@
+            requireNonNull(buildSymbols, "buildSymbols is null");
+            requireNonNull(types, "types is null");
+
+            buildTypes = dynamicFilter
```
Thanks, fixed.

rzeyde-varada commented:
Refactored and rebased (to resolve the merge conflict).

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,222 @@
+    private final OperatorContext context;
+    private boolean finished;
+    private Page current;
+    private final Consumer<TupleDomain<Integer>> dynamicPredicateConsumer;
```
Could you pass dynamic filter string id as tuple domain instead?

Channel index is specific to local execution and might be harder to interpret when we send dynamic filter to coordinator

rzeyde-varada replied:
Sounds good, refactored `LocalDynamicFilter` a bit to support that.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,222 @@
+            for (int position = 0; position < block.getPositionCount(); ++position) {
+                if (!valueSets[channelIndex].contains(block, position)) {
+                    valueSets[channelIndex].add(block, position);
+                    filterSizeInBytes += block.getEstimatedDataSizeForStats(position);
```
You can now use `io.prestosql.operator.aggregation.TypedSet#getRetainedSizeInBytes` after processing whole page. It should be considerable quicker and more accurate

rzeyde-varada replied:
Good catch, thanks!
https://github.com/prestosql/presto/blob/8d8f3e66d252d2b1275e639460d966d53f65977a/presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java#L162-L182
WDYT?

raunaqmorarka commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java:
```diff
@@ -0,0 +1,234 @@
+ * The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+ * We support only small build-side pages (which should be the case when using "broadcast" join).
+ */
+public class DynamicFilterCollectOperator
```
Should we rename this class to `DynamicFilterSourceOperator` ? It will help to avoid confusion with `DynamicFilterCollector` class that we'll need to add for the row filtering changes.

rzeyde-varada replied:
Sounds good, fixed and force-pushed https://github.com/prestosql/presto/pull/931/commits/4920e008fbb95337f42b984e6fee9c5d25301543.

rzeyde-varada commented:
Fixed a few issues found during TPCDS benchmarking.

raunaqmorarka commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+        }
+    }
+
+    public static class DynamicFilterCollectOperatorFactory
```
I think this should also be renamed to `DynamicFilterSourceOperatorFactory` now to be consistent

rzeyde-varada replied:
Good catch, thanks!
Fixed at fc8aef3.

rzeyde-varada commented:
Hive benchmarking results (using Presto 317 with Glue and S3) are attached here: [PDF](https://github.com/prestosql/presto/files/3519426/benchto.results.on.TPCDS.sf1000.-.chart.pdf)

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+    {
+        private final String filterId;
+        private final Type type;
+        private final int index;
```
nit: add `private final boolean closed`

rzeyde-varada replied:
Added to `DynamicFilterSourceOperatorFactory`.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+
+        @Override
+        public void noMoreOperators()
+        {
```
nit: set `closed = true`. See how other operators implement this method

rzeyde-varada replied:
Added (together with a `checkState`)

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            return new DynamicFilterSourceOperator(
```
nit: check state that operator is not closed. See how other operators implement this method

rzeyde-varada replied:
Thanks, added.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+    private final List<Channel> channels;
+
+    // May be dropped if the predicate becomes too large.
+    private BlockBuilder[] blockBuilders;
```
make both `@nullable`

rzeyde-varada replied:
Done.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+    public DynamicFilterSourceOperator(OperatorContext context, Consumer<TupleDomain<String>> dynamicPredicateConsumer, List<Channel> channels, PlanNodeId planNodeId)
+    {
+        this.context = requireNonNull(context, "context is null");
+        this.maxFilterPositionsCount = getDynamicFilteringMaxPerDriverValueCount(context.getSession());
```
calling this methods in constructor will affect benchmarking results as they parse strings. Pass these params to `DynamicFilterSourceOperatorFactory`.

rzeyde-varada replied:
Good catch, done.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+        this.dynamicPredicateConsumer = requireNonNull(dynamicPredicateConsumer, "dynamicPredicateConsumer is null");
+        this.channels = requireNonNull(channels, "channels is null");
+
+        final int expectedSize = 8;
```
make it static final field in class

rzeyde-varada replied:
Done.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+            Type type = channels.get(channelIndex).type;
+            this.blockBuilders[channelIndex] = type.createBlockBuilder(null, expectedSize);
+            this.valueSets[channelIndex] = new TypedSet(type, Optional.empty(), blockBuilders[channelIndex], expectedSize,
+                                                        String.format("DynamicFilterSourceOperator_%s_%d", planNodeId, channelIndex));
```
`TypedSet` has static mem limit. Currently it's `io.prestosql.operator.aggregation.TypedSet#MAX_FUNCTION_MEMORY`.
Either we limit max dynamic filter size or we make `TypedSet` configurable

rzeyde-varada replied:
Made `TypedSet` configurable.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+            for (int position = 0; position < block.getPositionCount(); ++position) {
+                if (!valueSet.contains(block, position)) {
+                    valueSet.add(block, position);
+                    filterPositionsCount += 1;
```
types set won't count for duplicates, so you shouldn't either. Please add method `TypedSet#positionCount` and use it for counting

rzeyde-varada replied:
IIUC, it can also be achieved by using the `TypedSet#size` count.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,234 @@
+                values.add(value);
+            }
+        }
+        // Inner-join doesn't match rows with null key column values.
```
`Inner and right join`

rzeyde-varada replied:
Done.

rzeyde-varada commented:
Thanks for the review!

raunaqmorarka commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,266 @@
+            }
+        }
+        catch (PrestoException e) {
+            if (e.getErrorCode().equals(EXCEEDED_FUNCTION_MEMORY_LIMIT.toErrorCode())) {
```
Wouldn't it be hard for users to figure out why dynamic filtering didn't apply with this approach ? How about we add an upper limit to DYNAMIC_FILTERING_MAX_PER_DRIVER_BYTE_SIZE in SystemSessionProperties ? There is already an upper bound of 1MB in FeatureConfig on `experimental.dynamic-filtering-max-per-driver-byte-size`

rzeyde-varada replied:
Following @sopel39 [suggestion](https://github.com/prestosql/presto/pull/931#discussion_r319014865), I am using a new `TypedSet` c-tor, which allows setting `maxBlockMemoryInBytes` to the specified session property:
https://github.com/prestosql/presto/blob/5fbf6338955ec70c3bc147b9059ab5157c7dc6f5/presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java#L153-L158
IIUC, it should throw `PrestoException` only if a single block builder exceeds the specified `maxBlockMemoryInBytes`.
What do you think?

raunaqmorarka replied:
I see, is the condition when exception is thrown essentially the same as `filterSizeInBytes > maxFilterSizeInBytes` ? I'm a bit confused by the presence of both the try/catch and the `filterSizeInBytes > maxFilterSizeInBytes` condition. If the TypedSet now already enforces maxFilterSizeInBytes by throwing exception, then maybe we get rid of the code around `filterSizeInBytes` ?

rzeyde-varada replied:
Good point, added a comment to make this clearer.

raunaqmorarka commented on presto-main/src/main/java/io/prestosql/SystemSessionProperties.java:
```diff
@@ -121,6 +121,8 @@
     public static final String QUERY_MAX_MEMORY_PER_NODE = "query_max_memory_per_node";
     public static final String QUERY_MAX_TOTAL_MEMORY_PER_NODE = "query_max_total_memory_per_node";
+    public static final String DYNAMIC_FILTERING_MAX_PER_DRIVER_VALUE_COUNT = "dynamic_filtering_max_per_driver_value_count";
+    public static final String DYNAMIC_FILTERING_MAX_PER_DRIVER_BYTE_SIZE = "dynamic_filtering_max_per_driver_byte_size";
```
Since the unit for this is DataSize now, `DYNAMIC_FILTERING_MAX_PER_DRIVER_SIZE = "dynamic_filtering_max_per_driver_size"` would be a more suitable name

rzeyde-varada replied:
Thanks, fixed.

rzeyde-varada replied:
Renamed a few more instances of `ByteSize` into `Size`.

raunaqmorarka commented on presto-main/src/main/java/io/prestosql/SystemSessionProperties.java:
```diff
@@ -523,7 +525,17 @@ public SystemSessionProperties(
+                        false),
+                dataSizeProperty(
+                        DYNAMIC_FILTERING_MAX_PER_DRIVER_BYTE_SIZE,
+                        "Experimental: maximum number of bytes to be collected for dynamic filtering per-driver",
```
Would be better to drop `bytes` from here as well

rzeyde-varada replied:
Thanks, fixed.

rzeyde-varada commented:
Thanks for the review!

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,272 @@
+                                                        blockBuilders[channelIndex],
+                                                        EXPECTED_BLOCK_BUILDER_SIZE,
+                                                        String.format("DynamicFilterSourceOperator_%s_%d", planNodeId, channelIndex),
+                                                        maxFilterSize);
```
I would simply make `TypedSet` unlimited here

rzeyde-varada replied:
Sounds good, thanks for the suggestion.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:
```diff
@@ -0,0 +1,211 @@
+        verify(probeSymbols.keySet().equals(buildChannels.keySet()), "probeSymbols and buildChannels must have same keys");
+
+        this.resultFuture = SettableFuture.create();
+        this.tupleDomainConsumer = this::addPartition;
```
not needed, inline into `getTupleDomainConsumer`

rzeyde-varada replied:
You're right, thanks!

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:
```diff
@@ -0,0 +1,211 @@
+    private final SettableFuture<TupleDomain<Symbol>> resultFuture;
+    private final Consumer<TupleDomain<String>> tupleDomainConsumer;
+
+    private TupleDomain<String>[] partitions;
```
make it `@Nullable`

rzeyde-varada replied:
Since we switched to use a single `TupleDomain`, I think that we can keep it non-`@Nullable`.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:
```diff
@@ -0,0 +1,211 @@
+    private final SettableFuture<TupleDomain<Symbol>> resultFuture;
+    private final Consumer<TupleDomain<String>> tupleDomainConsumer;
+
+    private TupleDomain<String>[] partitions;
```
just keep one `TupleDomain` which you union once partitions come by

rzeyde-varada replied:
Good catch, thanks!

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:
```diff
@@ -0,0 +1,211 @@
+
+    public static Optional<LocalDynamicFilter> create(JoinNode planNode, int partitionCount)
+    {
+        Set<String> localDynamicFilters = planNode.getDynamicFilters().keySet();
```
rename it to `joinDynamicFilters`

rzeyde-varada replied:
Done.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:
```diff
@@ -0,0 +1,211 @@
+                .entrySet()
+                .stream()
+                // Add build channels that match local probe symbols.
+                .filter(entry -> probeSymbols.containsKey(entry.getKey()))
```
`probeSymbols` are from probe side, right? However, `entry.getKey()` will be symbol from build side, so they will never match. I think you should filter by `localDynamicFilters`

rzeyde-varada replied:
I think that I got the comment wrong... instead it should say:
```
// Skip build channels that don't match local probe dynamic filters.        
```
The `probeSymbols` keys are actually probe-side dynamic filter IDs, so the `.filter(...)` above is meant to skip build-side dynamic filters which don't have matching probe-side symbols.
Please let me know if this can be rewritten in a simpler way :)

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:
```diff
@@ -0,0 +1,211 @@
+    private static final Logger log = Logger.get(LocalDynamicFilter.class);
+
+    // Mapping from dynamic filter ID to its probe symbols.
+    private final Multimap<String, Symbol> probeSymbols;
```
IIUC, in case there is a union on the probe side, we may need to match several probe-side symbols for a single dynamic filter.
For example, consider:
```
SELECT * 
FROM ((SELECT x FROM f1) UNION (SELECT x FROM f2)) f, d 
WHERE f.x = d.x AND d.y = 1; 
```
In this case, we should use the dynamic filter both for filtering `f1` and `f2`.

I've updated the comments to (hopefully) better explain this scenario. WDYT?

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java:
```diff
@@ -0,0 +1,211 @@
+    private static class Channel
+    {
+        private final int buildChannelIndex;
+        private final DynamicFilters.Descriptor descriptor;
```
Done.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,272 @@
+                handleTooLargePredicate();
+            }
+        }
+        catch (PrestoException e) {
```
Done.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,272 @@
+            for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+                Block block = page.getBlock(channels.get(channelIndex).index);
+                TypedSet valueSet = valueSets[channelIndex];
+                filterSizeInBytes -= valueSet.getRetainedSizeInBytes();
```
Good catch, thanks!

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,272 @@
+        for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+            Type type = channels.get(channelIndex).type;
+            this.blockBuilders[channelIndex] = type.createBlockBuilder(null, EXPECTED_BLOCK_BUILDER_SIZE);
+            this.valueSets[channelIndex] = new TypedSet(type,
```
Done.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,272 @@
+        for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+            Type type = channels.get(channelIndex).type;
+            this.blockBuilders[channelIndex] = type.createBlockBuilder(null, EXPECTED_BLOCK_BUILDER_SIZE);
+            this.valueSets[channelIndex] = new TypedSet(type,
```
Done.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    private static final int EXPECTED_BLOCK_BUILDER_SIZE = 8;
```
Done.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,272 @@
+
+        private boolean closed;
+
+        public DynamicFilterSourceOperatorFactory(int operatorId, PlanNodeId planNodeId, Consumer<TupleDomain<String>> dynamicPredicateConsumer, List<Channel> channels,
```
Done.

rzeyde-varada commented:
Thanks a lot for the review!
Fixed the issues, and force-pushed b0818dcbfa.

rzeyde-varada commented:
After getting a transient error in an assertion at `TestHiveDistributedJoinQueriesWithDynamicFiltering.testJoinWithEmptyBuildSide` (see [Travis log](https://api.travis-ci.com/v3/job/238727394/log.txt)), I've refactored operator's statistics retrieval a bit at 90de9b8ad4.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/aggregation/TypedSet.java:
```diff
@@ -46,14 +46,16 @@
     private final String functionName;
 
+    // Unlimited if empty.
+    private final Optional<Long> maxBlockMemoryInBytes;
```
You're right, fixed.

rzeyde-varada replied:
You're right, fixed.

rzeyde-varada commented on presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java:
```diff
@@ -0,0 +1,266 @@
+            return;  // the predicate became too large.
+        }
+
+        long filterSizeInBytes = 0;
```
Added.

rzeyde-varada commented:
Thanks!
Fixed and force-pushed b43519729b.

sopel39 commented:
@rzeyde-varada I've created a final PR with some comments addressed. Will merge it if there are no objections

https://github.com/prestosql/presto/pull/1686

rzeyde-varada commented:
Thanks a lot @sopel39!
#1686 looks great, I'd be happy to help with the follow-up [tasks](https://github.com/prestosql/presto/pull/1686#issuecomment-538977935).
